File: lecture2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    // Prompt the user for some text
    string text = get_string("Text: ");

    // Count the number of letters, words, and sentences in the text
    int letters = count_letters(text);

    int words = count_words(text);

    int sentences = count_sentences(text);

    // Compute the Coleman-Liau index

    int index = round(0.0588 * (float) letters * 100 / (float) words -
                      0.296 * (float) sentences * 100 / (float) words - 15.8);

    // Print the grade level
    if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

int count_letters(string text)
{
    // Return the number of letters in text
    int letters = 0;
    for (int i = 0, n = strlen(text); i <= n; i++)
    {
        int l = toupper(text[i]);
        if (l >= 65 && l <= 90)
        {
            letters++;
        }
    }
    return letters;
}

int count_words(string text)
{
    // Return the number of words in text
    int words = 1;
    for (int i = 0, n = strlen(text); i <= n; i++)
    {

        if (text[i] == ' ')
        {
            words++;
        }
    }
    return words;
}

int count_sentences(string text)
{
    // Return the number of sentences in text
    int sentences = 0;
    for (int i = 0, n = strlen(text); i <= n; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            if (text[i + 1] != '.' || text[i + 1] != '!' || text[i + 1] != '?')
            {
                sentences++;
            }
        }
        int l = toupper(text[i]);
    }
    if (sentences == 0)
    {
        for (int i = 0, n = strlen(text); i <= n; i++)
        {
            int l = toupper(text[i]);
            if (l >= 65 && l <= 90)
            {
                sentences = 1;
            }
        }
    }
    return sentences;
}

File: lecture6/sentimental-readability/readability.py
from cs50 import get_string
import re
from math import ceil


def count_letters(text):
    # Return the number of letters in text
    letters = sum(1 for char in text if char.isalpha())
    return letters


def count_words(text):
    # Split the text into words using whitespace as the delimiter
    words = text.split()
    # Return the number of words
    return len(words)


def count_sentences(text):
    # Return the number of sentences in text
    sentences = len(re.findall(r'[.!?]', text))
    return sentences


def main():
    # Prompt the user for some text
    text = get_string("Text: ")

    # Count the number of letters, words, and sentences in the text
    letters = count_letters(text)
    words = count_words(text)
    sentences = count_sentences(text)

    # Calculate the average number of letters and sentences per 100 words
    L = letters * 100 / words
    S = sentences * 100 / words

    # Compute the Coleman-Liau index
    index = round(0.0588 * L - 0.296 * S - 15.8)

    # Print the grade level
    if index > 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index}")


main()
